eth_p2p/rlpx_protocols/eth.nim

```diff
@@ -277,25 +329,124 @@ proc obtainBlocksFromPeer(peer: Peer, syncCtx: SyncContext) {.async.} =
           let b = await peer.getBlockBodies(hashes)
           bodies.add(b.get.blocks)
 
-        shallowCopy(workItem.bodies, bodies)
-        dataReceived = true
+        if bodies.len == workItem.headers.len:
+          shallowCopy(workItem.bodies, bodies)
+          dataReceived = true
+        else:
+          warn "Bodies len != headers.len", bodies = bodies.len, headers = workItem.headers.len
     except:
       # the success case uses `continue`, so we can just fall back to the
       # failure path below. If we signal time-outs with exceptions such
       # failures will be easier to handle.
       discard
 
     if dataReceived:
+      workItem.state = Received
       syncCtx.returnWorkItem workItemIdx
     else:
+      workItem.state = Initial
       try:
         await peer.disconnect(SubprotocolReason)
       except:
         discard
       syncCtx.handleLostPeer()
       break
 
-  debug "Nothing to sync"
+  debug "Fininshed otaining blocks", peer
+
+proc peersAgreeOnChain(a, b: Peer): Future[bool] {.async.} =
+  # Returns true if one of the peers acknowledges existense of the best block
+  # of another peer.
+  var
+    a = a
+    b = b
+
+  if a.state(eth).bestDifficulty < b.state(eth).bestDifficulty:
+    swap(a, b)
+
+  let request = BlocksRequest(
+    startBlock: HashOrNum(isHash: true,
+                          hash: b.state(eth).bestBlockHash),
+    maxResults: 1,
+    skip: 0,
+    reverse: true)
+
+  let latestBlock = await a.getBlockHeaders(request)
+  result = latestBlock.isSome and latestBlock.get.headers.len > 0
+
+proc randomTrustedPeer(ctx: SyncContext): Peer =
+  var k = rand(ctx.trustedPeers.len - 1)
+  var i = 0
+  for p in ctx.trustedPeers:
+    result = p
+    if i == k: return
+    inc i
+
+proc startSyncWithPeer(ctx: SyncContext, peer: Peer) {.async.} =
+  if ctx.trustedPeers.len >= minPeersToStartSync:
+    # We have enough trusted peers. Validate new peer against trusted
+    if await peersAgreeOnChain(peer, ctx.randomTrustedPeer()):
+      ctx.trustedPeers.incl(peer)
+      asyncCheck ctx.obtainBlocksFromPeer(peer)
+  elif ctx.trustedPeers.len == 0:
+    # Assume the peer is trusted, but don't start sync until we reevaluate
+    # it with more peers
+    debug "Assume trusted peer", peer
+    ctx.trustedPeers.incl(peer)
+  else:
+    # At this point we have some "trusted" candidates, but they are not
+    # "trusted" enough. We evaluate `peer` against all other candidates.
+    # If one of the candidates disagrees, we swap it for `peer`. If all
+    # candidates agree, we add `peer` to trusted set. The peers in the set
+    # will become "fully trusted" (and sync will start) when the set is big
+    # enough
+    var
+      agreeScore = 0
+      disagreedPeer: Peer
+
+    for tp in ctx.trustedPeers:
```

Comment:
> I can imagine using the current logic even as a potential DoS vector

Note that I assume there would be some way to distinguish incoming/outgoing peers, and in this case we want to consider only outgoing peers, which are kinda random and chosen by us. Also see #10.

Reply:
> This seems unnecessary, because if I have found to agree with the

I didn't think about it thoroughly (and likely I will add more details to this algorithm later), but here we must keep in mind what an "agreement" actually means. Agreement is an acknowledgement from one of the peers that another is on his chain. So it is theoretically possible that having agreement A-B and A-C you would have a disagreement C-B. So checking all the variants we effectively eliminate forked peers.